### ISSUE TITLE
Disallow single expression syntax when the function returns Unit

### ISSUE THREAD

**Issue by igorwojda** [detekt/detekt#1675]
OK
```
fun doSth() = getName()

fun getName() = "Igor"
```

Fail
```
fun doSth() = println("ABC")

fun println(): Unit {
    // ...
}
```

The main reason behind it is that it's not obvious that no value will be returned in 2nd example

**Comment by schalkms** [comment]
That's a good idea.

The following case is covered by the `OptionalUnit` rule.
```Kotlin
fun println(): Unit {
    // ...
}
```
This example isn't covered by the `OptionalUnit` rule. I propose to add that to the `OptionalUnit` rule. However, this requires type resolution.
```Kotlin
fun doSth() = println("ABC")
```

**Comment by igorwojda** [comment]
I just realized that we should do this for both `Unit` and `Nothing` types

**Comment by schalkms** [comment]
We shouldn't do this for the `Nothing` type btw.
It's perfectly valid to return `Nothing`.
Please take a look at the [Kotlin documentation](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-nothing.html).

> You can use Nothing to represent "a value that never exists": for example, if a function has the return type of Nothing, it means that it never returns (always throws an exception).

**Comment by schalkms** [comment]
So to conclude the rule should additionally handle functions where the return type needs to be resolved from the return expression.

```Kotlin
fun doSth() = println("ABC")
```

**Comment by igorwojda** [comment]
Sure, you can return `Northing` (sam as `Unit`), but the discussion here is not about "can it be done?", but rather the fact that using single expression function may lead to the wrong conclusion that function actually returns value. That's why we should to forbid expression body for functions return Unit/Nothing and make it. This way programmers will have to be more explicit about functions, not returning any "standard" value.

**Comment by schalkms** [comment]
> Sure, you can return `Northing` (sam as `Unit`), but the discussion here is not about "can it be done?", but rather the fact that using single expression function may lead to the wrong conclusion that function actually returns value. That's why we should to forbid expression body for functions return Unit/Nothing and make it. This way programmers will have to be more explicit about functions, not returning any "standard" value.

I agree that the discussion is not about whether it can be done. It's the responsibility of the compiler to do so.
As I mentioned, I also agree that this should be done with the `Unit` type. Hence, I also put the corresponding label on it.
I strongly disagree with the rest. The return type `Nothing` in Kotlin has a clear use case. (You can use Nothing to represent "a value that never exists")
There's no reason to forbid that. It's also no **style** related issue, since as a caller I clearly want to see that this function never returns.

If you have use cases to forbid the `Nothing` return type, then please use a [custom rule](https://arturbosch.github.io/detekt/extensions.html#custom-rulesets) for it.

**Comment by igorwojda** [comment]
I guess it's just a matter of perspective.

For me, if the function never returns value this single equality character `=` is a bit confusing (it look like an assignment but assignment actually never happens). The only two cases of `Nothing usages`: 
```
fun assertFalse(data: Boolean) = throw Exception("")

fun exit(data: Boolean) = thread.stop("")
```

THB I don't have strong preferences here and let' be honest `Nothing` is reserved for very specific cases, so it does not matter that much.

**Comment by schalkms** [comment]
That's why we leave it for now.
The following article also sheds some light on this topic.
https://proandroiddev.com/kotlins-nothing-type-946de7d464fb

You also wrote a fantastic comment with a valid use case. I'll quote your comment.

> There are more usages for Nothing in Kotlin language. One of them would be a function that does not return a value (do not confuse this with Unit where the function actually returns a value of type Unit). A typical example is an methodassertFail used for testing or method that exits current process. Both methods will never actually return any value yet we need to explicitly say tell it to a compiler using special type (Nothing).
```Kotlin
fun assertFail():Nothing {
   throw Exception()
}
```

**Comment by marschwar** [comment]
I do like the general idea behind this rule and would like to give it a shot if that is ok. 

Thinking about the issue I realized that it should be allowed to use the single expression syntax when the expression returns the `Unit` object and maybe even when specifying the return type explicitly. How about an expression that evaluates to `Unit?`? Should that be forbidden as well? Maybe `Unit?` should be forbidden in the first place ;)

I don't think the `OptionalUnit` rule would be such an ideal place to put the check. How about a `ForbiddenSingleExpressionSyntax` rule instead? 

**Non-compliant:**
```kotlin
fun test() = println("something")
fun test(param: String?) = param?.let { println("something") }
```

**Compliant:**
```kotlin
// There is no room for confusion here. This should be allowed or flagged in `EmptyFunctionBlock`.
fun test() = Unit

// This is ok (for this rule) since it does use single expression syntax
fun test() { }

// Return type specified - not sure if the rule should allow this
fun test(): Unit = println("something")
fun test(param: String?): Unit? = param?.let { println("something") }

// of course this is ok
fun test() {
    println("something")
}
```

**Comment by schalkms** [comment]
@marschwar please go ahead. Thank you for contributing! 😀
Please see my comments to the compliant code samples underneath.

```kotlin
// The rule OptionalUnit flags this.
fun test() = Unit

// The rule EmptyFunction flags this.
fun test() { }

// I agree with allowing this. The programmer made a deliberate choice here.
fun test(): Unit = println("something")
fun test(param: String?): Unit? = param?.let { println("something") }
```

**Comment by marschwar** [comment]
@igorwojda Coming back to your original issue. Please check if the rule `ImplicitUnitReturnType` is what you are looking for.

**Comment by schalkms** [comment]
```kotlin
fun doSth() = println("ABC")
```

`ImplicitUnitReturnType` flags this case.

```kotlin
fun println(): Unit {
    // ...
}
```

`OptionalUnit` of detekt and `NoUnitReturn` flag the second case.
So the behavior is consistent.

**Comment by schalkms** [comment]
@marschwar thanks for raising this case.
